kernel/src/transaction.rs

```diff
@@ -1,120 +1,249 @@
+        self.commit_info = Some(EngineCommitInfo {
+            data: commit_info.into(),
+            schema: schema.into(),
+        });
     }
 }
 
-#[cfg(all(test, feature = "default-engine"))]
-mod tests {
-    use std::path::PathBuf;
+/// Result after committing a transaction. If 'committed', the version is the new version written
+/// to the log. If 'conflict', the transaction is returned so the caller can resolve the conflict
+/// (along with the version which conflicted).
+pub enum CommitResult {
+    /// The transaction was successfully committed at the version.
+    Committed(crate::Version),
+    /// The transaction conflicted with an existing version (at the version given).
+    Conflict(Transaction, crate::Version),
+}
+
+// given the engine's commit info (data and schema as [EngineCommitInfo]) we want to create both:
+// (1) the commitInfo action to commit (and append more actions to) and
+// (2) the schema for the actions to commit (this is determined by the engine's commit info schema)
+#[allow(clippy::type_complexity)]
+fn generate_commit_info(
+    engine: &'_ dyn Engine,
+    engine_commit_info: Option<EngineCommitInfo>,
+) -> DeltaResult<(
+    Box<dyn Iterator<Item = Box<dyn EngineData>> + Send + '_>,
+    SchemaRef,
+)> {
+    use crate::actions::{
+        ADD_NAME, COMMIT_INFO_NAME, METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME, TRANSACTION_NAME,
+    };
+
+    // TODO: enforce single row commit info
+    // TODO: for now we always require commit info
+    let action_schema = Arc::new(engine_commit_info.as_ref().map_or(
+        get_log_schema().clone(),
+        |commit_info| {
+            let mut action_fields = get_log_schema().fields().collect::<Vec<_>>();
+            let commit_info_field = action_fields
+                .pop()
+                .expect("last field is commit_info in action schema");
+            let DataType::Struct(commit_info_schema) = commit_info_field.data_type() else {
+                unreachable!("commit_info_field is a struct");
+            };
+            let mut commit_info_fields = commit_info_schema.fields().collect::<Vec<_>>();
+            commit_info_fields.extend(commit_info.schema.fields());
+            let commit_info_schema =
+                StructType::new(commit_info_fields.into_iter().cloned().collect());
+            let mut action_fields = action_fields.into_iter().cloned().collect::<Vec<_>>();
+            action_fields.push(crate::schema::StructField::new(
+                COMMIT_INFO_NAME,
+                commit_info_schema,
+                true,
+            ));
+            StructType::new(action_fields)
+        },
+    ));
+
+    let action_schema_ref = Arc::clone(&action_schema);
+    let actions = engine_commit_info.into_iter().map(move |commit_info| {
+        let engine_commit_info_data = commit_info.data;
+        let engine_commit_info_schema = commit_info.schema;
+        let mut commit_info_expr = vec![Expression::literal(format!("v{}", KERNEL_VERSION))];
```

Comment:
Nit: May be nicer just sticking iwth iterators:
```rust
        let commit_info_expr = std::iter::once(Expression::literal(format!("v{}", KERNEL_VERSION)))
            .chain(
                engine_commit_info_schema
                    .fields()
                    .map(|f| Expression::column(f.name())),
            )
            .collect();
```

Reply:
yea this is nicer :)